extensions/android-paging3-coroutines/src/main/java/com/squareup/sqldelight/android/paging3/QueryPagingSource.kt

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.sqldelight.android.paging3
+
+import androidx.paging.PagingSource
+import androidx.paging.PagingState
+import com.squareup.sqldelight.Query
+import com.squareup.sqldelight.Transacter
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.withContext
+import kotlin.properties.Delegates
+
+class QueryPagingSource<RowType : Any>(
+  private val queryProvider: (limit: Long, offset: Long) -> Query<RowType>,
```

Comment:
Does this have to be offset-based? Limit+offset queries are generally inferior to keyed+limit queries that can use the index (and are also stable when inserts occur in the middle of the result set). It seems like we can remove the `loadSize * key` stuff internally and if someone wants to do offset-based pagination they can do that calculation themselves in the lambda (treating key as a page index).

Reply:
Or perhaps we can offer an extension function helper so you can continue to use a function reference with limit+offset queries. Something like `QueryPagingSource(limitOffsetQueryProvider(::query), ..)`

Reply:
I was also thinking that the `offset * key` stuff seemed dumb. I'm going to look into another solution.

Reply:
Can you give me an example of a query you're thinking about? I found [this](http://blog.ssokolow.com/archives/2009/12/23/sql-pagination-without-offset/) example, but don't know if that's exactly what you're thinking.

Reply:
Hm, does paging with key+index present problems when a consumer doesn't want to page incrementally? I'm thinking it's impossible to jump from page 1 to n without request each page in between. Limit+offset doesn't have that problem because it can just be calculated.

Reply:
we need offset based in cash unless we do a major refactor, which might make sense at some point but not now. Ideally we just support both, but given that we only support offset for existing paging i think its fine to only support that initially and then add keyed queries afterwards

Reply:
https://use-the-index-luke.com/sql/partial-results/fetch-next-page

https://blog.jooq.org/2013/10/26/faster-sql-paging-with-jooq-using-the-seek-method/ & https://blog.jooq.org/2013/11/18/faster-sql-pagination-with-keysets-continued/

Basically a `WHERE id > ? LIMIT 10` instead of `LIMIT 10 OFFSET ?`

Reply:
From the jooq article:

> Essentially, the Seek Method does not skip records before an OFFSET, but it skips records until the last record previously fetched. 

Maybe I'm misunderstanding something, but this confirms to me that seeking to page 5 from page 1 wouldn't be possible without enumerating every intermediate result set. I _think_ the paging library doesn't enforce sequential paging, and therefore I think this might be a problem.

Reply:
How would you feel about maintaining `OffsetQueryPagingSource` and `KeyedQueryPagingSource` separately as `internal` classes, and exposing a `fun QueryPagingSource(...)` factory function? That would avoid the need for something like `limitOffsetProvider(::keyedQuery, ...)` which I _think_ (based on [the above link](https://blog.jooq.org/2013/11/18/faster-sql-pagination-with-keysets-continued/)) would require more SQL queries to bridge the gap between the two. It seems like a bit of a burden for users who want to use offset based paging.

Reply:
Sounds good to me!

Reply:
After an offline discussion with @JakeWharton we've decided to defer the implementation of `KeyedQueryPagingSource`. We have, however, hidden the implementations behind a factory function to allow for a seamless addition later.